### ISSUE TITLE
Triplet Sampling 

### ISSUE THREAD

**Issue by Modexus** [microsoft/torchgeo#523]
I would like to implement the [Tile2Vec](https://arxiv.org/abs/1805.02855) paper using the torchgeo library.

Samples consist of triplets (anchor, positive, negative). The positive samples are chosen to be in a neighborhood based on distance, while the negative samples are outside of this neighborhood.

I have a working sampler TripletBatchGeoSampler derived from BatchGeoSampler, using a custom stack_triplet_samples collate function, returning a dictionary with "anchor", "positive", "negative" keys.

This is necessary as RasterDataset only allows to load one image as "image" (or "mask"). 

Is there another class better suited to this usecase?
Should RasterDataset be extended for this usecase or rather a new class introduced for multiple elements per sample?

**Comment by isaaccorley** [comment]
Instead of having a dict with "anchor", "positive", and "negative" keys you could return a batch of images of shape (B, N, C, H, W) where the N=3 representing the anchor/positive/negative for that sample.

**Comment by Modexus** [comment]
Thanks, that's true. 
Seems better as some contrastive methods don't use a negative pairing.

I am mostly wondering if there is a way to circumvent the usage of the specific collate function. 
Is there a way to directly sample a List (in this case 3 elements) and stack them into a batch of shape (B, 3, C, H, W)? without having select the corresponding elements in collate (currently using [::3])?

**Comment by isaaccorley** [comment]
I think all PyTorch's default collate fn does is use torch.stack so if your sampler returns a list (or maybe list of list) of 3 tensors it may end up automatically stacking them to (B, 3, C, H, W). It's worth a try at least.

I think this is the best way imo because you want to keep data in the batch form of images/masks so you can easily use augmentations and transforms

**Comment by Modexus** [comment]
If one yields a list (or list of list for batches) in the sampler the __getitem__ function from RasterDataset will receive a list of BoundingBoxes which will cause an error.

I tried two solutions both of which work.
The first is to write a specific collate_fn. However, this function needs to be aware of the batch_size and number of elements per sample (3 in case of triplet). 

The second is to adapt __getitem__ of RasterDataset for multiple elements. I achieved this by treating query as a list and stacking at the end. In case of a single element the stack is redundant and has to be squeezed.

I would be interested what approach would be suited for this library as triplet sampling (and more advanced versions than Tile2Vec) is interesting for self-supervised learning in remote sensing.
Maybe I am missing some obvious way.

**Comment by adamjstewart** [comment]
I've also been interested in reproducing Tile2Vec in TorchGeo. I think this discussion is going in the right direction. If I find any time over the summer I'll see what I can do to. If you succeed before me, please open a PR to add any datasets/triplet-samplers to TorchGeo!

**Comment by Modexus** [comment]
I have added a PR with the current implementation of the triplet sampler. It is a simple implementation of the neighborhood based sampling described in Tile2Vec.

I will implement Tile2Vec in the coming week(s) for torchgeo. For this reason I have created the PR for the whole Tile2Vec implementation for the possibility of collaboration.

Let me know if I should split this PR or not and about possible fixes/ideas #525 .

**Comment by Modexus** [comment]
Hey @calebrob6, I originally closed the PR as I had to change a lot for my own usage. I have tile2vec implemented but it is severely lacking in code standards and documentation. Hopefully, I can clean it up in the future and create a new PR.
